File: frontend/demo/component/contextmenu/context-menu-custom-item-data.ts
import 'Frontend/demo/init'; // hidden-source-line
import '@vaadin/context-menu';
import { html, LitElement } from 'lit';
import { customElement, state } from 'lit/decorators.js';
import type { ContextMenuItem, ContextMenuItemSelectedEvent } from '@vaadin/context-menu';
import { applyTheme } from 'Frontend/demo/theme';

@customElement('context-menu-custom-item-data')
export class Example extends LitElement {
  protected override createRenderRoot() {
    const root = super.createRenderRoot();
    applyTheme(root);
    return root;
  }

  // tag::snippet[]
  @state()
  private items: Array<ContextMenuItem<{ value: string }>> = [
    {
      text: 'Copy as plain text',
      value:
        'Context Menu\n\nContext Menu is a component that you can attach to any component to display a context menu.',
    },
    {
      text: 'Copy as HTML',
      value:
        '<h1>Context Menu</h1><p>Context Menu is a component that you can attach to any component to display a context menu.</p>',
    },
    {
      text: 'Copy as Markdown',
      value:
        '# Context Menu\n\nContext Menu is a component that you can attach to any component to display a context menu.',
    },
  ];

  protected override render() {
    return html`
      <vaadin-context-menu .items="${this.items}" @item-selected="${this.itemSelected}">
        <h1>Context Menu</h1>
        <p>
          Context Menu is a component that you can attach to any component to display a context
          menu.
        </p>
      </vaadin-context-menu>
    `;
  }

  itemSelected(e: ContextMenuItemSelectedEvent<ContextMenuItem<{ value?: string }>>) {
    const value = e.detail.value.value;
    if (value) {
      navigator.clipboard.writeText(value);
    }
  }

  // end::snippet[]
}

File: src/main/java/com/vaadin/demo/component/contextmenu/ContextMenuCustomItemData.java
package com.vaadin.demo.component.contextmenu;

import com.vaadin.demo.DemoExporter; // hidden-source-line
import com.vaadin.flow.component.contextmenu.ContextMenu;
import com.vaadin.flow.component.html.Div;
import com.vaadin.flow.component.html.H1;
import com.vaadin.flow.component.html.Paragraph;
import com.vaadin.flow.router.Route;

@Route("context-menu-custom-item-data")
public class ContextMenuCustomItemData extends Div {

    public ContextMenuCustomItemData() {
        // tag::snippet[]
        Div wrapper = new Div(new H1("Context Menu"), new Paragraph(
                "Menu Bar is a horizontal button bar with hierarchical drop-down menus."));

        ContextMenu menu = new ContextMenu();
        menu.setTarget(wrapper);

        menu.addItem("Copy as plain text", event -> {
            // Provide a custom value by adding a click listener that holds a
            // reference to that value
            String value = "Menu Bar\n\nMenu Bar is a horizontal button bar with hierarchical drop-down menus.";
            copyToClipboard(value);
        });

        menu.addItem("Copy as HTML", event -> {
            String value = "<h1>Menu Bar</h1><p>Menu Bar is a horizontal button bar with hierarchical drop-down menus.</p>";
            copyToClipboard(value);
        });

        menu.addItem("Copy as Markdown", event -> {
            String value = "# Menu Bar\n\nMenu Bar is a horizontal button bar with hierarchical drop-down menus.";
            copyToClipboard(value);
        });
        // end::snippet[]

        add(wrapper);
    }

    private void copyToClipboard(String value) {
        getUI().ifPresent(ui -> ui.getPage()
                .executeJs("window.navigator.clipboard.writeText($0);", value));
    }

    public static class Exporter // hidden-source-line
            extends DemoExporter<ContextMenuCustomItemData> { // hidden-source-line
    } // hidden-source-line
}
